firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/datagrams.txt

```diff
@@ -0,0 +1,446 @@
+
+  GIVEN("A Datagram receiver of buffer size 254 bytes and expected sequence number equal to 0xff")
+  // assertions carried out in GIVEN:
+  // The output_datagram's seq accessor method returns 0xff
+  // The value returned by the output_datagram's length accessor method is 8
+  // The payload buffer returned by the output_datagram's payload accessor method is '0xff 0x06 0xcd 0x6a 0xc2 0x7f 0xa1 0x5b'  
+
+    WHEN("transform is called on two parseable datagram bodies, whose headers have sequence numbers as 0 & 1 respectively")
+      THEN("The first transform method reports ok status")
+      THEN("After the first transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+      THEN("The second transform method reports ok status")
+      THEN("After the second transform method is called, the value returned by the output_datagram's seq accessor method is equal to the sequence field of the input_buffer body's header")
+      THEN("The value returned by the output_datagram's length accessor method is equal to the length field of the input_buffer body's header")
+      THEN("The paylaod returned by the output_datagram's paylaod accessor method is equal to the paylaod from the input_buffer body")
+      THEN("The output_datagram's payload accessor method returns a reference to the payload given in its constructor")
+      THEN("The data in the output_datagram's payload buffer is independent of the data in input_buffer")
+
+/* Datagram sender
+
+Scenario: Datagram Sender correctly generates datagram bodies from payloads
+  GIVEN("A Datagram sender of buffer size 254 bytes with next sequence equal to 0")
+    WHEN("A non empty payload buffer is given to the datagram sender")
+      THEN("The transform method reports ok status")
+      THEN("The seq field of the body's header is equal to the next sequence")
+      THEN("The length field of the body's header is equal to the size of the payload buffer given")
+      THEN("The body's payload section correctly stores the paylaod as '0xf9 0x23 0x4a 0xd4 0xe0'")
+      THEN("The output datagram is as expected '0x00 0x05 0xf9 0x23 0x4a 0xd4 0xe0'")
+
+    WHEN("The output datagram cannot hold enough data")
+      THEN("The transform method reports to invalid length")
+      THEN("The output buffer is as expected")
```

Comment:
And what's the sequence number of the next datagram it will generate? (I think it should still be 0). Also, what is the output buffer supposed to be? (I think it should remain unchanged from whatever it had before transform was called, whether it's empty or has data)

This applies to the next WHEN, too

Reply:
> And what's the sequence number of the next datagram it will generate? (I think it should still be 0)

next_seq is a private member, therefore we can't make any assertions on it.

Reply:
We can indirectly infer the value of next_seq by calling transform again in a THEN and checking the sequence number field of the resulting output_buffer. And ultimately we don't even care about the next_seq private member, we just care that the sequence number fields of our output_buffers increment on successful outputs and stay the same on errors

Reply:
Yes will do that.

Reply:
This case is redundant after adding `static_assert`. I'm leaving the comment unresolved if we need to discuss anything.